Date: Fri, 16 Sep 2016 20:54:00 -0400
From: Felix Janda <felix.janda@...teo.de>
To: musl@lists.openwall.com
Subject: [PATCH v2] add pthread_setname_np

the thread name is displayed by gdb's "info threads".
---
v1->v2: add _GNU_SOURCE, and "UL" to prctl arguments

I hope that there are no programs for which this function is
performance critical...

Thanks for the review so far!
---
 include/pthread.h               |  1 +
 src/thread/pthread_setname_np.c | 26 ++++++++++++++++++++++++++
 2 files changed, 27 insertions(+)
 create mode 100644 src/thread/pthread_setname_np.c

diff --git musl-1.1.5/include/pthread.h musl-fixed/include/pthread.h
index 3d2e0c4..94ef919 100644
--- musl-1.1.5/include/pthread.h
+++ musl-fixed/include/pthread.h
@@ -214,6 +214,7 @@ struct cpu_set_t;
 int pthread_getaffinity_np(pthread_t, size_t, struct cpu_set_t *);
 int pthread_setaffinity_np(pthread_t, size_t, const struct cpu_set_t *);
 int pthread_getattr_np(pthread_t, pthread_attr_t *);
+int pthread_setname_np(pthread_t, const char *);
 int pthread_tryjoin_np(pthread_t, void **);
 int pthread_timedjoin_np(pthread_t, void **, const struct timespec *);
 #endif
diff --git musl-1.1.5/src/thread/pthread_setname_np.c musl-fixed/src/thread/pthread_setname_np.c
new file mode 100644
index 0000000..82d35e1
--- /dev/null
+++ musl-fixed/src/thread/pthread_setname_np.c
@@ -0,0 +1,26 @@
+#define _GNU_SOURCE
+#include <fcntl.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/prctl.h>
+
+#include "pthread_impl.h"
+
+int pthread_setname_np(pthread_t thread, const char *name)
+{
+	int fd, cs, status = 0;
+	char f[sizeof "/proc/self/task//comm" + 3*sizeof(int)];
+	size_t len;
+
+	if ((len = strnlen(name, 16)) > 15) return ERANGE;
+
+	if (thread == pthread_self())
+		return prctl(PR_SET_NAME, (unsigned long)name, 0UL, 0UL, 0UL) ? errno : 0;
+
+	snprintf(f, sizeof f, "/proc/self/task/%d/comm", thread->tid);
+	pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cs);
+	if ((fd = open(f, O_WRONLY)) < 0 || write(fd, name, len) < 0) status = errno;
+	if (fd >= 0) close(fd);
+	pthread_setcancelstate(cs, 0);
+	return status;
+}
-- 
2.7.3
